File: RotationMatrixToEulerAngles/rotm2euler.cpp
/*
 * Copyright (c) 2016 Satya Mallick <spmallick@learnopencv.com>
 * All rights reserved. No warranty, explicit or implicit, provided.
 */

#include <opencv2/core.hpp>

#include <algorithm>
#include <array>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <limits>
#include <random>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

constexpr double kPi = 3.141592653589793238462643383279502884;
constexpr double kDefaultTolerance = 1e-9;

cv::Mat asDoubleRotationMatrix(const cv::Mat& input) {
    if (input.rows != 3 || input.cols != 3 || input.channels() != 1) {
        throw std::invalid_argument("rotation matrix must be single-channel and 3x3");
    }
    if (input.depth() != CV_32F && input.depth() != CV_64F) {
        throw std::invalid_argument("rotation matrix must use float32 or float64 values");
    }

    cv::Mat rotation;
    input.convertTo(rotation, CV_64F);
    for (int row = 0; row < rotation.rows; ++row) {
        for (int column = 0; column < rotation.cols; ++column) {
            if (!std::isfinite(rotation.at<double>(row, column))) {
                throw std::invalid_argument("rotation matrix must contain finite values");
            }
        }
    }
    return rotation;
}

bool isRotationMatrix(const cv::Mat& input, const double tolerance = kDefaultTolerance) {
    if (!(tolerance > 0.0)) {
        throw std::invalid_argument("tolerance must be positive");
    }

    cv::Mat rotation;
    try {
        rotation = asDoubleRotationMatrix(input);
    } catch (const std::invalid_argument&) {
        return false;
    }

    const cv::Mat identity = cv::Mat::eye(3, 3, CV_64F);
    const cv::Mat shouldBeIdentity = rotation.t() * rotation;
    const double identityError = cv::norm(identity - shouldBeIdentity, cv::NORM_INF);
    const double determinantError = std::abs(cv::determinant(rotation) - 1.0);
    return identityError <= tolerance && determinantError <= tolerance;
}

cv::Mat eulerAnglesToRotationMatrix(const cv::Vec3d& theta) {
    for (int index = 0; index < 3; ++index) {
        if (!std::isfinite(theta[index])) {
            throw std::invalid_argument("Euler angles must contain finite values");
        }
    }

    const double sinX = std::sin(theta[0]);
    const double cosX = std::cos(theta[0]);
    const double sinY = std::sin(theta[1]);
    const double cosY = std::cos(theta[1]);
    const double sinZ = std::sin(theta[2]);
    const double cosZ = std::cos(theta[2]);

    const cv::Mat rotationX = cv::Mat(cv::Matx33d(
        1.0, 0.0, 0.0,
        0.0, cosX, -sinX,
        0.0, sinX, cosX));
    const cv::Mat rotationY = cv::Mat(cv::Matx33d(
        cosY, 0.0, sinY,
        0.0, 1.0, 0.0,
        -sinY, 0.0, cosY));
    const cv::Mat rotationZ = cv::Mat(cv::Matx33d(
        cosZ, -sinZ, 0.0,
        sinZ, cosZ, 0.0,
        0.0, 0.0, 1.0));
    return rotationZ * rotationY * rotationX;
}

cv::Vec3d rotationMatrixToEulerAngles(
    const cv::Mat& input,
    const double validationTolerance = kDefaultTolerance,
    const double singularEpsilon = kDefaultTolerance) {
    const cv::Mat rotation = asDoubleRotationMatrix(input);
    if (!isRotationMatrix(rotation, validationTolerance)) {
        throw std::invalid_argument("input is not a proper rotation matrix");
    }
    if (!(singularEpsilon > 0.0)) {
        throw std::invalid_argument("singular epsilon must be positive");
    }

    const double horizontalNorm =
        std::hypot(rotation.at<double>(0, 0), rotation.at<double>(1, 0));
    const bool singular = horizontalNorm < singularEpsilon;

    double xAngle = 0.0;
    double yAngle = 0.0;
    double zAngle = 0.0;
    if (!singular) {
        xAngle = std::atan2(rotation.at<double>(2, 1), rotation.at<double>(2, 2));
        yAngle = std::atan2(-rotation.at<double>(2, 0), horizontalNorm);
        zAngle = std::atan2(rotation.at<double>(1, 0), rotation.at<double>(0, 0));
    } else {
        xAngle = std::atan2(-rotation.at<double>(1, 2), rotation.at<double>(1, 1));
        yAngle = std::atan2(-rotation.at<double>(2, 0), horizontalNorm);
        zAngle = 0.0;
    }
    return cv::Vec3d(xAngle, yAngle, zAngle);
}

double matrixRoundTripError(const cv::Vec3d& angles) {
    const cv::Mat original = eulerAnglesToRotationMatrix(angles);
    const cv::Vec3d recoveredAngles = rotationMatrixToEulerAngles(original);
    const cv::Mat recovered = eulerAnglesToRotationMatrix(recoveredAngles);
    return cv::norm(original - recovered, cv::NORM_INF);
}

void require(const bool condition, const std::string& message) {
    if (!condition) {
        throw std::runtime_error(message);
    }
}

void runValidation() {
    const std::array<cv::Vec3d, 6> fixedCases = {
        cv::Vec3d(0.0, 0.0, 0.0),
        cv::Vec3d(kPi / 2.0, 0.0, 0.0),
        cv::Vec3d(0.0, kPi / 2.0, 0.0),
        cv::Vec3d(0.0, -kPi / 2.0, 0.0),
        cv::Vec3d(0.3, -0.7, 1.2),
        cv::Vec3d(-2.4, kPi / 2.0 - 1e-10, 0.8),
    };

    double maximumError = 0.0;
    for (const cv::Vec3d& angles : fixedCases) {
        maximumError = std::max(maximumError, matrixRoundTripError(angles));
    }

    std::mt19937 generator(20260726U);
    std::uniform_real_distribution<double> angleDistribution(-kPi, kPi);
    constexpr int kRandomCaseCount = 512;
    for (int index = 0; index < kRandomCaseCount; ++index) {
        const cv::Vec3d angles(
            angleDistribution(generator),
            angleDistribution(generator),
            angleDistribution(generator));
        maximumError = std::max(maximumError, matrixRoundTripError(angles));
    }

    const cv::Mat reflection = cv::Mat(cv::Matx33d(
        1.0, 0.0, 0.0,
        0.0, 1.0, 0.0,
        0.0, 0.0, -1.0));
    require(!isRotationMatrix(reflection), "reflection matrix was incorrectly accepted");
    require(maximumError <= 1e-9, "matrix round-trip error exceeded tolerance");

    const int totalCases = static_cast<int>(fixedCases.size()) + kRandomCaseCount;
    std::cout << "VALIDATION PASSED: " << totalCases
              << " cases, max_matrix_error=" << std::scientific << maximumError << '\n';
}

double parseDouble(const std::string& text) {
    std::size_t consumed = 0;
    const double value = std::stod(text, &consumed);
    if (consumed != text.size() || !std::isfinite(value)) {
        throw std::invalid_argument("invalid finite number: " + text);
    }
    return value;
}

struct Options {
    bool validate = false;
    bool degrees = false;
    bool help = false;
    bool hasAngles = false;
    bool hasMatrix = false;
    cv::Vec3d angles;
    cv::Mat matrix;
};

Options parseArguments(const int argc, char** argv) {
    Options options;
    for (int index = 1; index < argc; ++index) {
        const std::string argument(argv[index]);
        if (argument == "--validate") {
            options.validate = true;
        } else if (argument == "--degrees") {
            options.degrees = true;
        } else if (argument == "--help" || argument == "-h") {
            options.help = true;
        } else if (argument == "--angles") {
            if (index + 3 >= argc) {
                throw std::invalid_argument("--angles requires X Y Z");
            }
            const double xAngle = parseDouble(argv[++index]);
            const double yAngle = parseDouble(argv[++index]);
            const double zAngle = parseDouble(argv[++index]);
            options.angles = cv::Vec3d(xAngle, yAngle, zAngle);
            options.hasAngles = true;
        } else if (argument == "--matrix") {
            if (index + 9 >= argc) {
                throw std::invalid_argument("--matrix requires nine row-major values");
            }
            options.matrix = cv::Mat(3, 3, CV_64F);
            for (int element = 0; element < 9; ++element) {
                options.matrix.at<double>(element / 3, element % 3) =
                    parseDouble(argv[++index]);
            }
            options.hasMatrix = true;
        } else {
            throw std::invalid_argument("unknown argument: " + argument);
        }
    }

    const int selectedModes = static_cast<int>(options.validate) +
                              static_cast<int>(options.hasAngles) +
                              static_cast<int>(options.hasMatrix);
    if (selectedModes > 1) {
        throw std::invalid_argument("choose only one of --validate, --angles, or --matrix");
    }
    return options;
}

void printUsage(const char* program) {
    std::cout
        << "Usage: " << program << " [--degrees] [--angles X Y Z | --matrix R00 ... R22 | --validate]\n"
        << "Convention: active right-handed column-vector rotation R = Rz @ Ry @ Rx.\n";
}

}  // namespace

int main(const int argc, char** argv) {
    try {
        const Options options = parseArguments(argc, argv);
        if (options.help) {
            printUsage(argv[0]);
            return 0;
        }
        if (options.validate) {
            runValidation();
            return 0;
        }

        cv::Vec3d angles;
        cv::Mat rotation;
        if (options.hasMatrix) {
            rotation = asDoubleRotationMatrix(options.matrix);
            angles = rotationMatrixToEulerAngles(rotation);
        } else {
            angles = options.hasAngles ? options.angles : cv::Vec3d(20.0, -35.0, 70.0);
            if (!options.hasAngles || options.degrees) {
                angles *= kPi / 180.0;
            }
            rotation = eulerAnglesToRotationMatrix(angles);
        }

        const cv::Vec3d displayedAngles =
            options.degrees ? angles * (180.0 / kPi) : angles;
        std::cout << std::setprecision(10);
        std::cout << "Euler angles [x, y, z] ("
                  << (options.degrees ? "degrees" : "radians") << "):\n"
                  << displayedAngles << "\n";
        std::cout << "Rotation matrix Rz * Ry * Rx:\n" << rotation << "\n";

        const cv::Vec3d recoveredAngles = rotationMatrixToEulerAngles(rotation);
        const cv::Mat recovered = eulerAnglesToRotationMatrix(recoveredAngles);
        const double error = cv::norm(rotation - recovered, cv::NORM_INF);
        std::cout << "Round-trip matrix infinity-norm error: "
                  << std::scientific << error << '\n';
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "Error: " << error.what() << '\n';
        printUsage(argv[0]);
        return 2;
    }
}

File: RotationMatrixToEulerAngles/rotm2euler.py
#!/usr/bin/env python3
"""Convert between XYZ Tait-Bryan angles and 3x3 rotation matrices.

The convention used throughout this example is an active, right-handed
rotation of column vectors:

    R = Rz(z) @ Ry(y) @ Rx(x)

The returned angle vector is ordered ``[x, y, z]`` and is expressed in radians.
Euler/Tait-Bryan angles are not unique, so round-trip validation compares
rotation matrices rather than comparing angle vectors directly.
"""

from __future__ import annotations

import argparse
import math
from collections.abc import Sequence

import numpy as np
import numpy.typing as npt

FloatArray = npt.NDArray[np.float64]


def _as_finite_matrix(matrix: npt.ArrayLike) -> FloatArray:
    """Return *matrix* as a finite 3x3 float64 array or raise ValueError."""

    array = np.asarray(matrix, dtype=np.float64)
    if array.shape != (3, 3):
        raise ValueError(f"rotation matrix must have shape (3, 3), got {array.shape}")
    if not np.all(np.isfinite(array)):
        raise ValueError("rotation matrix must contain only finite values")
    return array


def is_rotation_matrix(matrix: npt.ArrayLike, tolerance: float = 1e-9) -> bool:
    """Return True when *matrix* is a proper 3D rotation matrix.

    A proper rotation must be orthonormal and have determinant +1. Checking the
    determinant rejects reflections, which satisfy ``R.T @ R == I`` but are not
    rotations.
    """

    if tolerance <= 0:
        raise ValueError("tolerance must be positive")
    try:
        rotation = _as_finite_matrix(matrix)
    except (TypeError, ValueError):
        return False

    identity_error = np.linalg.norm(
        rotation.T @ rotation - np.eye(3, dtype=np.float64), ord=np.inf
    )
    determinant_error = abs(float(np.linalg.det(rotation)) - 1.0)
    return identity_error <= tolerance and determinant_error <= tolerance


def euler_angles_to_rotation_matrix(theta: npt.ArrayLike) -> FloatArray:
    """Build ``Rz(z) @ Ry(y) @ Rx(x)`` from ``[x, y, z]`` radians."""

    angles = np.asarray(theta, dtype=np.float64)
    if angles.shape != (3,):
        raise ValueError(f"theta must have shape (3,), got {angles.shape}")
    if not np.all(np.isfinite(angles)):
        raise ValueError("theta must contain only finite values")

    x_angle, y_angle, z_angle = (float(value) for value in angles)
    sin_x, cos_x = math.sin(x_angle), math.cos(x_angle)
    sin_y, cos_y = math.sin(y_angle), math.cos(y_angle)
    sin_z, cos_z = math.sin(z_angle), math.cos(z_angle)

    rotation_x = np.array(
        [[1.0, 0.0, 0.0], [0.0, cos_x, -sin_x], [0.0, sin_x, cos_x]],
        dtype=np.float64,
    )
    rotation_y = np.array(
        [[cos_y, 0.0, sin_y], [0.0, 1.0, 0.0], [-sin_y, 0.0, cos_y]],
        dtype=np.float64,
    )
    rotation_z = np.array(
        [[cos_z, -sin_z, 0.0], [sin_z, cos_z, 0.0], [0.0, 0.0, 1.0]],
        dtype=np.float64,
    )
    return rotation_z @ rotation_y @ rotation_x


def rotation_matrix_to_euler_angles(
    matrix: npt.ArrayLike,
    *,
    validation_tolerance: float = 1e-9,
    singular_epsilon: float = 1e-9,
) -> FloatArray:
    """Return one valid ``[x, y, z]`` representation for a rotation matrix.

    At gimbal lock, infinitely many angle triples represent the same rotation.
    This implementation chooses ``z = 0`` and solves for ``x`` and ``y``.
    """

    rotation = _as_finite_matrix(matrix)
    if not is_rotation_matrix(rotation, validation_tolerance):
        raise ValueError("input is not a proper rotation matrix")
    if singular_epsilon <= 0:
        raise ValueError("singular_epsilon must be positive")

    horizontal_norm = math.hypot(float(rotation[0, 0]), float(rotation[1, 0]))
    singular = horizontal_norm < singular_epsilon

    if not singular:
        x_angle = math.atan2(float(rotation[2, 1]), float(rotation[2, 2]))
        y_angle = math.atan2(-float(rotation[2, 0]), horizontal_norm)
        z_angle = math.atan2(float(rotation[1, 0]), float(rotation[0, 0]))
    else:
        x_angle = math.atan2(-float(rotation[1, 2]), float(rotation[1, 1]))
        y_angle = math.atan2(-float(rotation[2, 0]), horizontal_norm)
        z_angle = 0.0

    return np.array([x_angle, y_angle, z_angle], dtype=np.float64)


# Compatibility aliases keep existing tutorial imports working.
isRotationMatrix = is_rotation_matrix
eulerAnglesToRotationMatrix = euler_angles_to_rotation_matrix
rotationMatrixToEulerAngles = rotation_matrix_to_euler_angles


def _matrix_round_trip_error(theta: npt.ArrayLike) -> float:
    """Return the infinity-norm error after an angles/matrix round trip."""

    original = euler_angles_to_rotation_matrix(theta)
    recovered_angles = rotation_matrix_to_euler_angles(original)
    recovered = euler_angles_to_rotation_matrix(recovered_angles)
    return float(np.linalg.norm(original - recovered, ord=np.inf))


def run_validation() -> dict[str, float | int]:
    """Run deterministic regression checks used by tests and CTest parity."""

    cases = [
        np.zeros(3, dtype=np.float64),
        np.array([math.pi / 2.0, 0.0, 0.0]),
        np.array([0.0, math.pi / 2.0, 0.0]),
        np.array([0.0, -math.pi / 2.0, 0.0]),
        np.array([0.3, -0.7, 1.2]),
        np.array([-2.4, math.pi / 2.0 - 1e-10, 0.8]),
    ]

    random_generator = np.random.default_rng(20260726)
    random_cases = random_generator.uniform(-math.pi, math.pi, size=(512, 3))
    errors = [_matrix_round_trip_error(case) for case in cases]
    errors.extend(_matrix_round_trip_error(case) for case in random_cases)
    maximum_error = max(errors)

    reflection = np.diag([1.0, 1.0, -1.0])
    if is_rotation_matrix(reflection):
        raise AssertionError("reflection matrix was incorrectly accepted")
    if maximum_error > 1e-9:
        raise AssertionError(f"round-trip error {maximum_error:.3e} exceeded tolerance")

    result: dict[str, float | int] = {
        "cases": len(errors),
        "max_matrix_error": maximum_error,
    }
    print(
        "VALIDATION PASSED: "
        f"{result['cases']} cases, max_matrix_error={maximum_error:.3e}"
    )
    return result


def _build_parser() -> argparse.ArgumentParser:
    """Create the command-line parser."""

    parser = argparse.ArgumentParser(
        description="Convert XYZ Tait-Bryan angles and 3x3 rotation matrices."
    )
    inputs = parser.add_mutually_exclusive_group()
    inputs.add_argument(
        "--angles",
        nargs=3,
        type=float,
        metavar=("X", "Y", "Z"),
        help="angles ordered X Y Z; radians unless --degrees is present",
    )
    inputs.add_argument(
        "--matrix",
        nargs=9,
        type=float,
        metavar=("R00", "R01", "R02", "R10", "R11", "R12", "R20", "R21", "R22"),
        help="row-major 3x3 rotation matrix",
    )
    inputs.add_argument(
        "--validate",
        action="store_true",
        help="run deterministic regression checks",
    )
    parser.add_argument(
        "--degrees",
        action="store_true",
        help="read and print angles in degrees instead of radians",
    )
    return parser


def main(argv: Sequence[str] | None = None) -> int:
    """Run the deterministic command-line demonstration."""

    arguments = _build_parser().parse_args(argv)
    if arguments.validate:
        run_validation()
        return 0

    if arguments.matrix is not None:
        rotation = np.asarray(arguments.matrix, dtype=np.float64).reshape(3, 3)
        angles = rotation_matrix_to_euler_angles(rotation)
    else:
        supplied = arguments.angles or [20.0, -35.0, 70.0]
        angles = np.asarray(supplied, dtype=np.float64)
        if arguments.angles is None or arguments.degrees:
            angles = np.deg2rad(angles)
        rotation = euler_angles_to_rotation_matrix(angles)

    displayed_angles = np.rad2deg(angles) if arguments.degrees else angles
    units = "degrees" if arguments.degrees else "radians"
    print(f"Euler angles [x, y, z] ({units}):")
    print(np.array2string(displayed_angles, precision=10, suppress_small=True))
    print("Rotation matrix Rz @ Ry @ Rx:")
    print(np.array2string(rotation, precision=10, suppress_small=True))

    recovered = rotation_matrix_to_euler_angles(rotation)
    recovered_rotation = euler_angles_to_rotation_matrix(recovered)
    error = float(np.linalg.norm(rotation - recovered_rotation, ord=np.inf))
    print(f"Round-trip matrix infinity-norm error: {error:.3e}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
